File: agent/tools/marketing/posthog.py
"""PostHog product-analytics read tools for the marketing department.

Read-only. The agent uses these to answer founder questions like
"how many signups this week?" or "which event is firing most?". No
event tracking writes — those happen client-side in app code.

Auth: IntegrationConnection (toolkit='posthog') → POSTHOG_API_KEY env.
Project id from IntegrationConnection.accessKey or POSTHOG_PROJECT_ID.
Host from IntegrationConnection.metadata.host or POSTHOG_HOST
(default https://us.posthog.com).
"""

from __future__ import annotations

import os
from typing import Any

import httpx
import structlog

from agents import function_tool
from db import supabase

logger = structlog.get_logger(__name__)

_DEFAULT_HOST = "https://us.posthog.com"
_TIMEOUT = 20.0
_MAX_DAYS = 365
_MAX_LIMIT = 50


async def _get_posthog_config(space_id: str) -> tuple[str | None, str | None, str]:
    """Return (api_key, project_id, host). api_key/project_id may be None."""
    api_key: str | None = None
    project_id: str | None = None
    host = _DEFAULT_HOST

    if space_id:
        try:
            db = await supabase()
            res = await (
                db.table("IntegrationConnection")
                .select("accessToken, accessKey, metadata")
                .eq("spaceId", space_id)
                .eq("toolkit", "posthog")
                .eq("status", "active")
                .maybe_single()
                .execute()
            )
            if res.data:
                api_key = res.data.get("accessToken") or None
                project_id = res.data.get("accessKey") or None
                meta = res.data.get("metadata") or {}
                if isinstance(meta, dict) and meta.get("host"):
                    host = str(meta["host"])
        except Exception as err:  # noqa: BLE001
            logger.warning("posthog_config_lookup_failed", space_id=space_id, error=str(err)[:200])

    api_key = api_key or os.environ.get("POSTHOG_API_KEY")
    project_id = project_id or os.environ.get("POSTHOG_PROJECT_ID")
    host = os.environ.get("POSTHOG_HOST", host).rstrip("/")
    return api_key, project_id, host


def _auth_headers(key: str) -> dict[str, str]:
    return {
        "Authorization": f"Bearer {key}",
        "Content-Type": "application/json",
    }


def _clamp(value: int, lo: int, hi: int) -> int:
    if value < lo:
        return lo
    if value > hi:
        return hi
    return value


def _escape(s: str) -> str:
    return s.replace("'", "''")


async def _hogql(
    client: httpx.AsyncClient,
    host: str,
    project_id: str,
    api_key: str,
    query: str,
) -> tuple[list[list[Any]] | None, str | None]:
    """POST to /api/projects/{id}/query/. Returns (results, error)."""
    resp = await client.post(
        f"{host}/api/projects/{project_id}/query/",
        headers=_auth_headers(api_key),
        json={"query": {"kind": "HogQLQuery", "query": query}},
    )
    if resp.status_code == 401:
        return None, "PostHog auth failed. Check API key or reconnect PostHog."
    if resp.status_code == 403:
        return None, "PostHog forbidden — the personal API key lacks project access."
    if resp.status_code == 429:
        return None, "PostHog rate limit hit. Retry shortly."
    if not resp.is_success:
        body: dict[str, Any] = {}
        try:
            body = resp.json()
        except Exception:  # noqa: BLE001
            pass
        msg = body.get("detail") or body.get("error") or resp.text[:200]
        return None, f"PostHog error {resp.status_code}: {msg}"

    data = resp.json()
    return data.get("results") or [], None


@function_tool
async def posthog_event_count(
    event_name: str,
    days: int = 30,
    space_id: str = "",
) -> str:
    """Return the count of an event over the last N days."""
    api_key, project_id, host = await _get_posthog_config(space_id)
    if not api_key or not project_id:
        return "No PostHog credentials found. Connect PostHog in settings or set POSTHOG_API_KEY + POSTHOG_PROJECT_ID."

    d = _clamp(days, 1, _MAX_DAYS)
    q = (
        f"SELECT count() FROM events WHERE event = '{_escape(event_name)}' "
        f"AND timestamp >= now() - INTERVAL {d} DAY"
    )

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        results, err = await _hogql(client, host, project_id, api_key, q)
    if err:
        return err
    assert results is not None
    n = 0
    if results and results[0]:
        try:
            n = int(results[0][0])
        except (TypeError, ValueError):
            n = 0
    return f"{event_name}: {n} events in the last {d} days."


@function_tool
async def posthog_top_events(
    days: int = 30,
    limit: int = 10,
    space_id: str = "",
) -> str:
    """Return the most-fired events with their counts."""
    api_key, project_id, host = await _get_posthog_config(space_id)
    if not api_key or not project_id:
        return "No PostHog credentials found. Connect PostHog in settings or set POSTHOG_API_KEY + POSTHOG_PROJECT_ID."

    d = _clamp(days, 1, _MAX_DAYS)
    lim = _clamp(limit, 1, _MAX_LIMIT)
    q = (
        f"SELECT event, count() AS c FROM events "
        f"WHERE timestamp >= now() - INTERVAL {d} DAY "
        f"GROUP BY event ORDER BY c DESC LIMIT {lim}"
    )

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        results, err = await _hogql(client, host, project_id, api_key, q)
    if err:
        return err
    assert results is not None
    if not results:
        return f"No events recorded in the last {d} days."
    lines = [f"Top events (last {d} days):"]
    for row in results:
        ev = str(row[0]) if row and len(row) > 0 else "(unknown)"
        try:
            c = int(row[1]) if row and len(row) > 1 else 0
        except (TypeError, ValueError):
            c = 0
        lines.append(f"  {ev}: {c}")
    return "\n".join(lines)


@function_tool
async def posthog_active_users(days: int = 30, space_id: str = "") -> str:
    """Return the count of unique active distinct_ids over the last N days."""
    api_key, project_id, host = await _get_posthog_config(space_id)
    if not api_key or not project_id:
        return "No PostHog credentials found. Connect PostHog in settings or set POSTHOG_API_KEY + POSTHOG_PROJECT_ID."

    d = _clamp(days, 1, _MAX_DAYS)
    q = (
        f"SELECT count(DISTINCT distinct_id) FROM events "
        f"WHERE timestamp >= now() - INTERVAL {d} DAY"
    )

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        results, err = await _hogql(client, host, project_id, api_key, q)
    if err:
        return err
    assert results is not None
    n = 0
    if results and results[0]:
        try:
            n = int(results[0][0])
        except (TypeError, ValueError):
            n = 0
    return f"{n} active users in the last {d} days."


@function_tool
async def posthog_funnel(
    steps_csv: str,
    days: int = 30,
    space_id: str = "",
) -> str:
    """Compute step-by-step conversion through a sequence of events.

    steps_csv: comma-separated list of event names, in order.
    """
    api_key, project_id, host = await _get_posthog_config(space_id)
    if not api_key or not project_id:
        return "No PostHog credentials found. Connect PostHog in settings or set POSTHOG_API_KEY + POSTHOG_PROJECT_ID."

    steps = [s.strip() for s in steps_csv.split(",") if s.strip()]
    if not steps:
        return "Funnel needs at least one event name (comma-separated)."

    d = _clamp(days, 1, _MAX_DAYS)
    counts: list[int] = []

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        for ev in steps:
            q = (
                f"SELECT count(DISTINCT distinct_id) FROM events "
                f"WHERE event = '{_escape(ev)}' "
                f"AND timestamp >= now() - INTERVAL {d} DAY"
            )
            results, err = await _hogql(client, host, project_id, api_key, q)
            if err:
                return err
            assert results is not None
            n = 0
            if results and results[0]:
                try:
                    n = int(results[0][0])
                except (TypeError, ValueError):
                    n = 0
            counts.append(n)

    lines = [f"Funnel (last {d} days):"]
    for i, ev in enumerate(steps):
        c = counts[i]
        if i == 0:
            lines.append(f"  1. {ev}: {c}")
        else:
            prev = counts[i - 1]
            pct = (c / prev * 100.0) if prev > 0 else 0.0
            lines.append(f"  {i + 1}. {ev}: {c} ({pct:.1f}% of previous)")
    return "\n".join(lines)


@function_tool
async def posthog_signups_trend(days: int = 30, space_id: str = "") -> str:
    """Return the daily count of 'user signed up' events for the last N days."""
    api_key, project_id, host = await _get_posthog_config(space_id)
    if not api_key or not project_id:
        return "No PostHog credentials found. Connect PostHog in settings or set POSTHOG_API_KEY + POSTHOG_PROJECT_ID."

    d = _clamp(days, 1, _MAX_DAYS)
    q = (
        f"SELECT toDate(timestamp) AS day, count() AS c FROM events "
        f"WHERE event = 'user signed up' "
        f"AND timestamp >= now() - INTERVAL {d} DAY "
        f"GROUP BY day ORDER BY day ASC"
    )

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        results, err = await _hogql(client, host, project_id, api_key, q)
    if err:
        return err
    assert results is not None
    if not results:
        return f"No signups recorded in the last {d} days."
    lines = [f"Signups by day (last {d} days):"]
    total = 0
    for row in results:
        day = str(row[0]) if row and len(row) > 0 else "?"
        try:
            c = int(row[1]) if row and len(row) > 1 else 0
        except (TypeError, ValueError):
            c = 0
        total += c
        lines.append(f"  {day}: {c}")
    lines.append(f"Total: {total}")
    return "\n".join(lines)

File: lib/integrations/adapters/posthog.ts
/**
 * Thin PostHog REST API adapter — read-only.
 *
 * Charles uses this to answer founder questions like "how many signups
 * this week?" without shipping a chart library. Event ingestion happens
 * client-side in app code; this adapter only queries.
 *
 * Auth priority:
 *   1. IntegrationConnection (toolkit='posthog', status='active') —
 *      `accessToken` is the personal API key, `accessKey` is the project id,
 *      `metadata.host` is the region host.
 *   2. process.env.POSTHOG_API_KEY / POSTHOG_PROJECT_ID / POSTHOG_HOST.
 *
 * Throws when API key or project id can't be resolved.
 */

import { supabase } from '@/lib/supabase';

const DEFAULT_HOST = 'https://us.posthog.com';

export interface PostHogConfig {
  host: string;
  apiKey: string;
  projectId: string;
}

export interface EventCount {
  event: string;
  count: number;
}

export interface TrendPoint {
  day: string;
  count: number;
}

export interface FunnelStep {
  event: string;
  count: number;
  conversionFromPrevious: number;
}

interface PostHogConnectionRow {
  accessToken?: string | null;
  accessKey?: string | null;
  metadata?: { host?: string | null } | null;
}

export async function posthogGetConfig(spaceId: string): Promise<PostHogConfig> {
  const { data } = await supabase
    .from('IntegrationConnection')
    .select('accessToken, accessKey, metadata')
    .eq('spaceId', spaceId)
    .eq('toolkit', 'posthog')
    .eq('status', 'active')
    .maybeSingle();

  const row = (data as PostHogConnectionRow | null) ?? null;

  const apiKey = row?.accessToken || process.env.POSTHOG_API_KEY;
  const projectId = row?.accessKey || process.env.POSTHOG_PROJECT_ID;
  const host = row?.metadata?.host || process.env.POSTHOG_HOST || DEFAULT_HOST;

  if (!apiKey || !projectId) {
    throw new Error(
      `PostHog: no API key or project id found for space ${spaceId}. ` +
        `Connect PostHog in Settings → Integrations, or set POSTHOG_API_KEY and POSTHOG_PROJECT_ID.`,
    );
  }

  return { host: host.replace(/\/$/, ''), apiKey, projectId };
}

function authHeaders(apiKey: string): HeadersInit {
  return {
    Authorization: `Bearer ${apiKey}`,
    'Content-Type': 'application/json',
  };
}

async function readOrThrow(res: Response, action: string): Promise<unknown> {
  if (!res.ok) {
    let body: { detail?: string; error?: string; message?: string } = {};
    try {
      body = (await res.json()) as { detail?: string; error?: string; message?: string };
    } catch {
      // ignore
    }
    const msg = body.detail ?? body.error ?? body.message ?? (await res.text().catch(() => ''));
    throw new Error(`PostHog ${action} failed: ${res.status} — ${msg}`);
  }
  return res.json();
}

function clampDays(days: number | undefined, fallback = 30): number {
  const d = Math.floor(days ?? fallback);
  if (!Number.isFinite(d) || d < 1) return 1;
  if (d > 365) return 365;
  return d;
}

function clampLimit(limit: number | undefined, fallback = 10): number {
  const l = Math.floor(limit ?? fallback);
  if (!Number.isFinite(l) || l < 1) return 1;
  if (l > 50) return 50;
  return l;
}

async function hogql(
  cfg: PostHogConfig,
  query: string,
  action: string,
): Promise<{ columns: string[]; results: unknown[][] }> {
  // Uses POST /api/projects/:id/query/ with kind=HogQLQuery — simplest
  // surface to count + group rows without the trends-insight machinery.
  const res = await fetch(`${cfg.host}/api/projects/${cfg.projectId}/query/`, {
    method: 'POST',
    headers: authHeaders(cfg.apiKey),
    body: JSON.stringify({ query: { kind: 'HogQLQuery', query } }),
  });
  const data = (await readOrThrow(res, action)) as {
    columns?: string[];
    results?: unknown[][];
  };
  return { columns: data.columns ?? [], results: data.results ?? [] };
}

// ── Operations ───────────────────────────────────────────────────────────────

// HogQL: cheaper than POST /insights/trend for a single scalar.
export async function posthogEventCount(
  spaceId: string,
  event: string,
  days = 30,
): Promise<number> {
  const cfg = await posthogGetConfig(spaceId);
  const d = clampDays(days);
  const safeEvent = event.replace(/'/g, "''");
  const q =
    `SELECT count() FROM events WHERE event = '${safeEvent}' ` +
    `AND timestamp >= now() - INTERVAL ${d} DAY`;
  const { results } = await hogql(cfg, q, 'eventCount');
  const row = results[0];
  const n = row && typeof row[0] === 'number' ? row[0] : 0;
  return n;
}

// HogQL: GROUP BY event ORDER BY count — cleaner than the trends endpoint
// for an unbounded set of events.
export async function posthogTopEvents(
  spaceId: string,
  opts: { days?: number; limit?: number } = {},
): Promise<EventCount[]> {
  const cfg = await posthogGetConfig(spaceId);
  const d = clampDays(opts.days);
  const l = clampLimit(opts.limit);
  const q =
    `SELECT event, count() AS c FROM events ` +
    `WHERE timestamp >= now() - INTERVAL ${d} DAY ` +
    `GROUP BY event ORDER BY c DESC LIMIT ${l}`;
  const { results } = await hogql(cfg, q, 'topEvents');
  return results.map((row) => ({
    event: String(row[0] ?? ''),
    count: typeof row[1] === 'number' ? row[1] : Number(row[1] ?? 0),
  }));
}

// HogQL: count(distinct distinct_id) — one scalar, no trend series.
export async function posthogActiveUsers(spaceId: string, days = 30): Promise<number> {
  const cfg = await posthogGetConfig(spaceId);
  const d = clampDays(days);
  const q =
    `SELECT count(DISTINCT distinct_id) FROM events ` +
    `WHERE timestamp >= now() - INTERVAL ${d} DAY`;
  const { results } = await hogql(cfg, q, 'activeUsers');
  const row = results[0];
  return row && typeof row[0] === 'number' ? row[0] : 0;
}

// HogQL: bucket per day so the dashboard can render a tidy sparkline
// without parsing a trends response shape.
export async function posthogSignupsTrend(
  spaceId: string,
  days = 30,
): Promise<TrendPoint[]> {
  const cfg = await posthogGetConfig(spaceId);
  const d = clampDays(days);
  const q =
    `SELECT toDate(timestamp) AS day, count() AS c FROM events ` +
    `WHERE event = 'user signed up' ` +
    `AND timestamp >= now() - INTERVAL ${d} DAY ` +
    `GROUP BY day ORDER BY day ASC`;
  const { results } = await hogql(cfg, q, 'signupsTrend');
  return results.map((row) => ({
    day: String(row[0] ?? ''),
    count: typeof row[1] === 'number' ? row[1] : Number(row[1] ?? 0),
  }));
}

// HogQL: per-step counts then derive conversion in code — the funnel insight
// endpoint adds breakdowns we don't need.
export async function posthogFunnel(
  spaceId: string,
  events: string[],
  days = 30,
): Promise<FunnelStep[]> {
  const cfg = await posthogGetConfig(spaceId);
  const clean = events.map((e) => e.trim()).filter(Boolean);
  if (clean.length === 0) return [];
  const d = clampDays(days);

  const counts: number[] = [];
  for (const ev of clean) {
    const safe = ev.replace(/'/g, "''");
    const q =
      `SELECT count(DISTINCT distinct_id) FROM events ` +
      `WHERE event = '${safe}' ` +
      `AND timestamp >= now() - INTERVAL ${d} DAY`;
    const { results } = await hogql(cfg, q, `funnel step "${ev}"`);
    const row = results[0];
    counts.push(row && typeof row[0] === 'number' ? row[0] : 0);
  }

  return clean.map((event, i) => {
    const count = counts[i] ?? 0;
    const prev = i === 0 ? count : counts[i - 1] ?? 0;
    const conversion = i === 0 ? 1 : prev > 0 ? count / prev : 0;
    return { event, count, conversionFromPrevious: conversion };
  });
}
